acvm-repo/acvm/src/compiler/optimizers/mod.rs

```diff
@@ -42,20 +53,22 @@ pub(super) fn optimize_internal(acir: Circuit) -> (Circuit, Vec<usize>) {
         .collect();
     let acir = Circuit { opcodes, ..acir };
 
-    // Track original acir opcode positions throughout the transformation passes of the compilation
-    // by applying the modifications done to the circuit opcodes and also to the opcode_positions (delete and insert)
-    let acir_opcode_positions = (0..acir.opcodes.len()).collect();
-
     // Unused memory optimization pass
     let memory_optimizer = UnusedMemoryOptimizer::new(acir);
     let (acir, acir_opcode_positions) =
         memory_optimizer.remove_unused_memory_initializations(acir_opcode_positions);
 
+    let (acir, acir_opcode_positions) =
```

Comment:
Did you find improvements from running the back-propagate pass  twice rather than simply at the end?

Reply:
The backpropagation pass  can get blocked where it determines that a witness is being asserted to be a constant but won't remove the witness as it's being used in a range constraint. The redundant range pass removes the range constraints from these witnesses so it unblocks the backpropagation pass.

I found that running the backpropagation pass twice allowed us to unpick a lot of these whereas just running it once at the end leaves these extra witnesses in.

Reply:
Maybe add a comment here explaining that rationale but this is nit.